### DISCUSSION TITLE
Disable the micrometer metrics which are enabled by default

### DISCUSSION THREAD

**Discussion by ShubhamRwt** [quarkusio/quarkus#32721] [Q&A] [answered]
Hi, 

I tried to disable the `quarkus.micrometer.enabled` which disables the micrometer metrics support completely and I was able to do that but then I got errors for `MetricsReporter` class and `RestBridge` class 

```java
Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type io.micrometer.prometheus.PrometheusMeterRegistry and qualifiers [@Default]
        - java member: io.strimzi.kafka.bridge.quarkus.MetricsReporter#meterRegistry
        - declared on CLASS bean [types=[io.strimzi.kafka.bridge.quarkus.MetricsReporter, java.lang.Object], qualifiers=[@Default, @Any], target=io.strimzi.kafka.bridge.quarkus.MetricsReporter]
```
 which I think are reasonable since I think that disabling the property removes the classes required

Is there some way I can disable the micrometer metrics or maybe a way to disable/ignore the endpoint `/q/metrics`?

[a comment by an automated account is not shown]

**Reply by brunobat** [reply]
Can you please provide a list of dependencies or a reproducer?
This feature has this test: io.quarkus.micrometer.deployment.MicrometerDisabledTest and it seems to be working.

**Reply by ShubhamRwt** [reply]
We are using ` quarkus-micrometer-registry-prometheus` dependency in our pom file. We use the micrometer related classes in the [MetricsReporter](https://github.com/ppatierno/strimzi-kafka-bridge/blob/add-metrics-quarkus/src/main/java/io/strimzi/kafka/bridge/quarkus/MetricsReporter.java#L26) class. Then we are injecting this class into [RestBridge](https://github.com/ppatierno/strimzi-kafka-bridge/blob/add-metrics-quarkus/src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java#L89) class where we are setting the metrics on `/metrics` endpoint https://github.com/ppatierno/strimzi-kafka-bridge/blob/add-metrics-quarkus/src/main/java/io/strimzi/kafka/bridge/quarkus/RestBridge.java#L350 . When we run the project then the micrometer meter metrics are there on both `/metrics` and `/q/metrics` but `/q/metrics` is not eaxctly useful to us. So I was trying if we can disable it. The mechanism we have introduced in order to get the metrics on `/metrics` endpoint that part give error when we disable the micrometer metrics.

**Reply by ppatierno** [reply]
@brunobat your test is right and it matches with what we have. By disabling via `quarkus.micrometer.enabled=false`, the `MeterRegistry` is not injected at all.
I think what Shubham is asking is about a way to disable metrics at "runtime" as usual :-) We had this same conversation about OpenTelemetry but still even in this case `quarkus.micrometer.enabled` seems to be "locked" at build time.
So what's the best workaround to disable metrics on `/q/metrics`? or forwarding requests on it to our `/metrics` maybe?

**Reply by brunobat** [reply]
Yes, You can now (Quarkus 3.0) disable OpenTelemetry at runtime but not Micrometer.
You must build 2 instances of the service, one with Micrometer support and another one without.
In relation to the Injection, you can do:
```
 @Inject
 Instance<MetricsReporter> metricsReporter;
 ```
 And check if it's there...

**Reply by ppatierno** [reply]
@brunobat yeah I know about your plan about OTel, we discussed that. Regarding building two instances I don't think it's a feasible way to solve the problem. I think Quarkus should follow the same approach as OTel, providing a way to disable stuff at runtime tbh.

**Reply by brunobat** [reply]
Micrometer is not OTel. 
Will evaluate your request.

**Reply by ppatierno** [reply]
> Micrometer is not OTel.

Yeah I know that. I was just making the point about "fixing" (enabled/disabled) stuff at build time and not allowing doing at runtime. I think it's valid for any feature.

**Reply by ebullient** [reply]
you can disable publishing of any MeterRegistry which will make it a no-op.. But the enable/disable is specifically to avoid bytecode manipulation and build-time injection, etc..

**Reply by ebullient** [reply, marked as answer]
You don't need to use the prometheus meter registry. You can (as I've shown) make your own.

Use the micrometer extension alone, and provide your own Micrometer MeterRegistry that does what you need and provides its own endpoint, or try what I've suggested on another discussion thread to replace the default prometheus meter registry so you can alter the scrape behavior that the default metrics handler invokes.

**Reply by ShubhamRwt** [reply]
@ebullient I am trying the way you suggested. I was confused at some step so left a comment there. Just saw that you have put a thumbs up there

**Reply by andreiyusupau** [reply]
@ebullient could you please share the link to example?
